File: USACO/1.1/beads/beads.c
/* 
ID: aiden.b1
TASK: beads
LANG: C
*/

#include <stdio.h>
#include <stdbool.h>

#define DEBUG 1
#ifdef DEBUG
#define DEBUG_PRINT printf 
#else
#define DEBUG_PRINT(...) 
#endif

int getNumBeads(char *beads, int size, int loc){
    //go forward
    DEBUG_PRINT("size %d  loc %d \n", size, loc);
    int fwd = 1;
    char last = beads[loc];
    int i = 1; 
    int nextInd = (loc + i) % size;
    while( last == beads[nextInd] ) {
        DEBUG_PRINT("next ind %d \t fwd %d \t \n",nextInd, fwd);
        nextInd = (loc + i) % size;
        fwd++; 
        i++;
    }

    int bw = 1;
    i = 1;
    last = beads[loc - 1];
    nextInd = (loc - i) % size;
    while( last == beads[nextInd] ) {
        DEBUG_PRINT("next ind %d \t bw %d \t \n",nextInd, bw);
        nextInd = (loc - i) % size;
        bw++; 
        i++;
    }


    return bw + fwd;
}


int main(void){
    FILE *fin = fopen("beads.in","r");
    FILE *fout = fopen("beads.out","w");

    int n ;
    fscanf(fin,"%d",&n);
    char beads[n];
    fscanf(fin,"%s",beads);
    
    int max = 0;
    int maxLoc = 0; 

    for(int i = 0; i<n; i++){
        int curr = getNumBeads(beads,n,i);
        
        if( curr > max){
            max = curr;
            maxLoc = i;
        }
    }


    printf("%d", maxLoc);    
    return 0;
}

File: problems/usaco/1.2/beads.py
"""
ID: aiden.b1
LANG: PYTHON3
TASK: beads
"""
fin = open ('beads.in', 'r')
fout = open ('beads.out', 'w')

fin.readline()
beads = fin.readline()[0:-1]

i = 0
last = 'a'
max_run = 0
while i < len(beads):
    curr_run = 0
    k = 0
    color = beads[i]
    while k < len(beads):
        ind = (i + k) % len(beads)
        if color == 'w' or beads[ind] == color or beads[ind] == 'w':
            if beads[ind] != 'w':
                color = beads[ind]
            curr_run += 1
            k += 1
        else:
            break

    k = 0
    color = beads[i - 1]
    while k < len(beads):
        ind = (i - k - 1) % len(beads)
        if color == 'w' or beads[ind] == color or beads[ind] == 'w':
            if beads[ind] != 'w':
                color = beads[ind]
            curr_run += 1
            k += 1
        else:
            break

    if curr_run > max_run:
        max_run = curr_run
    i += 1

fout.write(str(min(max_run, len(beads))) + "\n")
